src/basic/fileio.c

```diff
@@ -1076,35 +1076,66 @@ int fdopen_independent(int fd, const char *mode, FILE **ret) {
         return 0;
 }
 
-static int search_and_fopen_internal(
+static int search_and_open_internal(
                 const char *path,
-                const char *mode,
+                int mode,            /* if ret_fd is NULL this is an F_OK/X_OK/R_OK/W_OK mode for access(), otherwise an open mode for open() */
                 const char *root,
                 char **search,
-                FILE **ret,
+                int *ret_fd,
                 char **ret_path) {
 
+        int r;
+
         assert(path);
-        assert(mode);
-        assert(ret);
+
+        if (path_is_absolute(path)) {
+                _cleanup_close_ int fd = -EBADF;
+                bool success;
+
+                if (ret_fd) {
+                        fd = open(path, mode, 0777);
+                        success = fd >= 0;
+                } else
+                        success = access(path, mode) >= 0;
```

Comment:
This seems messy, how about this:

```
if (ret_fd) {
        fd = r = RET_NERRNO(open(path, mode, 0777));
else
        r = RET_NERRNO(access(path, mode));
if (r < 0)
        return r; 
```

You can technically even drop the `RET_NERRNO`